man/git-comment.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GIT-COMMENT 1"
.TH GIT-COMMENT 1 "2015-06-21" "1.0.0.0" "Git Comment Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
.Vb 1
\&    git\-comment \- Add comments to git commits
.Ve
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 5
\&    git comment [\-m <msg>] [\-\-amend <comment>] [\-c <commit>] [<filepath:line>]
\&    git comment \-\-delete <comment>
\&    git comment \-\-configure\-remote <remote>
\&    git comment \-\-help
\&    git comment \-\-version
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
git-comment adds, removes, or amends comments attached to git commits, without changing the commits themselves.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-\-amend <comment>" 4
.IX Item "--amend <comment>"
Edit an existing comment. No modifications are made if a commit is specified but no comment with a given \s-1ID\s0 attached to the given commit.
.IP "\fI\-d\fR <comment>" 4
.IX Item "-d <comment>"
Remove a comment
.IP "\fI\-h\fR" 4
.IX Item "-h"
Gives a pretty-printed usage of the command
.IP "\fI\-m\fR <msg>, \-\-message=<msg>" 4
.IX Item "-m <msg>, --message=<msg>"
Use the provided message instead of prompting for message content
.IP "\-\-version" 4
.IX Item "--version"
Print the current version number
.IP "\-\-configure\-remote" 4
.IX Item "--configure-remote"
Update the remote configuration to fetch and push comment changes by default
.IP "\fI\-c\fR <commit>, \-\-commit <commit>" 4
.IX Item "-c <commit>, --commit <commit>"
A commit
.IP "<filepath:line>" 4
.IX Item "<filepath:line>"
A reference to a file and line number, to make the comment more specific
.IP "<remote>" 4
.IX Item "<remote>"
A remote name
.SH "ENVIRONMENT AND CONFIGURATION"
.IX Header "ENVIRONMENT AND CONFIGURATION"
git-comment uses the same editor, author, and pager settings as git. See
\&\fIgit-var\fR(1) for more details.
.SH "HOOKS"
.IX Header "HOOKS"
This command can run pre-comment and post-comment hooks.
.SH "DISCUSSION"
.IX Header "DISCUSSION"
Comment objects are blobs containing extra information about a commit and optionally a changed line within a the commit.
.SH "AUTHOR"
.IX Header "AUTHOR"
git-comment was written and is maintained by Delisa Mason <delisam@acm.org>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIgit-comment-log\fR(1), \fIgit-comment-grep\fR(1), \fIgit-var\fR(1)
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2015 Delisa Mason <delisam@acm.org>
All rights reserved.


src/git-comment.go
package main

import (
	"errors"
	"fmt"
	gitc "git_comment"
	goopt "github.com/droundy/goopt"
	"os"
)

var buildVersion string
var message = goopt.String([]string{"-m", "--message"}, "", "comment message")
var amendID = goopt.String([]string{"--amend"}, "", "ID of a comment to amend. `--message` is required")
var deleteID = goopt.String([]string{"--delete"}, "", "ID of a comment to delete")
var printVersion = goopt.Flag([]string{"-v", "--version"}, []string{}, "Show the version number", "")

func main() {
	goopt.Parse(nil)
	var err error
	pwd, osErr := os.Getwd()
	handleError(osErr)
	if len(goopt.Args) > 2 {
		handleError(errors.New("Too many arguments provided"))
	} else if *printVersion {
		fmt.Println(buildVersion)
	} else if len(*deleteID) > 0 {
		err = gitc.DeleteComment(pwd, *deleteID)
		handleError(err)
	} else if len(*message) > 0 {
		var commit *string = nil
		var fileref = ""
		if len(goopt.Args) > 1 {
			fileref = goopt.Args[1]
		}
		if len(goopt.Args) > 0 {
			commit = &goopt.Args[0]
		}
		if len(*amendID) > 0 {
			err = gitc.UpdateComment(pwd, *amendID, *message)
			handleError(err)
		} else {
			id, err := gitc.CreateComment(pwd, commit, gitc.CreateFileRef(fileref), *message)
			handleError(err)
			fmt.Printf("[%v] Comment created", (*id)[:7])
		}
	} else {
		fmt.Println(goopt.Help())
	}
}

func handleError(err error) {
	if err != nil {
		fmt.Println(err)
		fmt.Println(goopt.Help())
		os.Exit(1)
	}
}
